File: runtime/values.go
package runtime

import (
	"fmt"
	"strings"

	"novago/frontend"
)

type ValueType string

const (
	NullValType     ValueType = "null"
	NumberValType   ValueType = "number"
	BooleanValType  ValueType = "boolean"
	ObjectValType   ValueType = "object"
	NativeFnValType ValueType = "native-fn"
	FunctionValType ValueType = "function"
	StringValType   ValueType = "string"
	ArrayValType    ValueType = "array"
	PromiseValType  ValueType = "promise"
)

type RuntimeVal interface {
	GetType() ValueType
}

type PromiseVal struct {
	Type   ValueType
	Result chan RuntimeVal
}

func (p *PromiseVal) GetType() ValueType { return p.Type }
func MK_PROMISE() *PromiseVal {
	return &PromiseVal{Type: PromiseValType, Result: make(chan RuntimeVal, 1)}
}

type NullVal struct {
	Type ValueType
}

func (n *NullVal) GetType() ValueType { return n.Type }
func MK_NULL() *NullVal {
	return &NullVal{Type: NullValType}
}

type BooleanVal struct {
	Type  ValueType
	Value bool
}

func (b *BooleanVal) GetType() ValueType { return b.Type }
func MK_BOOL(b bool) *BooleanVal {
	return &BooleanVal{Type: BooleanValType, Value: b}
}

type NumberVal struct {
	Type  ValueType
	Value float64
}

func (n *NumberVal) GetType() ValueType { return n.Type }
func MK_NUMBER(n float64) *NumberVal {
	return &NumberVal{Type: NumberValType, Value: n}
}

type StringVal struct {
	Type  ValueType
	Value string
}

func (s *StringVal) GetType() ValueType { return s.Type }
func MK_STRING(s string) *StringVal {
	return &StringVal{Type: StringValType, Value: s}
}

type ObjectVal struct {
	Type       ValueType
	Properties map[string]RuntimeVal
}

func (o *ObjectVal) GetType() ValueType { return o.Type }
func MK_OBJECT(props map[string]RuntimeVal) *ObjectVal {
	return &ObjectVal{Type: ObjectValType, Properties: props}
}

type ArrayVal struct {
	Type     ValueType
	Elements []RuntimeVal
}

func (a *ArrayVal) GetType() ValueType { return a.Type }
func MK_ARRAY(elements []RuntimeVal) *ArrayVal {
	return &ArrayVal{Type: ArrayValType, Elements: elements}
}

type FunctionCall func(args []RuntimeVal, env *Environment) RuntimeVal

type NativeFnVal struct {
	Type       ValueType
	Call       FunctionCall
	Properties map[string]RuntimeVal
}

func (n *NativeFnVal) GetType() ValueType { return n.Type }
func MK_NATIVE_FN(call FunctionCall) *NativeFnVal {
	return &NativeFnVal{Type: NativeFnValType, Call: call, Properties: make(map[string]RuntimeVal)}
}

type FunctionVal struct {
	Type           ValueType
	Name           string
	Parameters     []string
	DeclarationEnv *Environment
	Body           []frontend.Stmt
	Async          bool
}

func (f *FunctionVal) GetType() ValueType { return f.Type }

type ReturnException struct {
	Value RuntimeVal
}

func (e *ReturnException) Error() string {
	return "Return"
}

func PlainStringify(val RuntimeVal) string {
	switch v := val.(type) {
	case *StringVal:
		return v.Value
	case *NumberVal:
		return fmt.Sprintf("%v", v.Value)
	case *BooleanVal:
		return fmt.Sprintf("%v", v.Value)
	case *NullVal:
		return "null"
	case *ArrayVal:
		var strs []string
		for _, e := range v.Elements {
			strs = append(strs, PlainStringify(e))
		}
		return "[" + strings.Join(strs, ", ") + "]"
	case *ObjectVal:
		var strs []string
		for k, val := range v.Properties {
			strs = append(strs, fmt.Sprintf("%s: %s", k, PlainStringify(val)))
		}
		return "{ " + strings.Join(strs, ", ") + " }"
	case *FunctionVal:
		return "[Function]"
	case *NativeFnVal:
		return "[Native Function]"
	default:
		return fmt.Sprintf("%v", v)
	}
}

func ToNumber(val RuntimeVal) float64 {
	switch v := val.(type) {
	case *NumberVal:
		return v.Value
	case *StringVal:
		var f float64
		fmt.Sscanf(v.Value, "%f", &f)
		return f
	case *BooleanVal:
		if v.Value {
			return 1
		}
		return 0
	default:
		return 0
	}
}

func ToBool(val RuntimeVal) bool {
	switch v := val.(type) {
	case *BooleanVal:
		return v.Value
	case *NumberVal:
		return v.Value != 0
	case *StringVal:
		return v.Value != ""
	case *NullVal:
		return false
	default:
		return true
	}
}

func GoToRuntimeVal(goVal interface{}) RuntimeVal {
	if goVal == nil {
		return MK_NULL()
	}

	switch v := goVal.(type) {
	case float64:
		return MK_NUMBER(v)
	case int:
		return MK_NUMBER(float64(v))
	case string:
		return MK_STRING(v)
	case bool:
		return MK_BOOL(v)
	case []interface{}:
		elements := make([]RuntimeVal, 0, len(v))
		for _, item := range v {
			elements = append(elements, GoToRuntimeVal(item))
		}
		return MK_ARRAY(elements)
	case map[string]interface{}:
		props := make(map[string]RuntimeVal)
		for key, item := range v {
			props[key] = GoToRuntimeVal(item)
		}
		return MK_OBJECT(props)
	default:
		return MK_NULL()
	}
}

File: novascript/src/runtime/values.ts

// src/runtime/values.ts

import { Statement } from "../frontend/ast";
import Environment from "./environment";

export type ValueType = "null" | "number" | "boolean" | "object" | "native-fn" | "function" | "string" | "array" | "promise";

export interface RuntimeVal {
  type: ValueType;
  underlyingValue?: any; // For native JS bridging
}

export interface PromiseVal extends RuntimeVal {
    type: "promise";
    promise: Promise<RuntimeVal>;
}

export interface NullVal extends RuntimeVal {
  type: "null";
  value: null;
}

export interface BooleanVal extends RuntimeVal {
  type: "boolean";
  value: boolean;
}

export interface NumberVal extends RuntimeVal {
  type: "number";
  value: number;
}

export interface StringVal extends RuntimeVal {
    type: "string";
    value: string;
}

export interface ObjectVal extends RuntimeVal {
  type: "object";
  properties: Map<string, RuntimeVal>;
}

export interface ArrayVal extends RuntimeVal {
    type: "array";
    elements: RuntimeVal[];
}

export class ReturnException extends Error {
    value: RuntimeVal;
    constructor(value: RuntimeVal) {
        super("Return");
        this.value = value;
    }
}

export type FunctionCall = (args: RuntimeVal[], env: Environment) => RuntimeVal | Promise<RuntimeVal>;

export interface NativeFnVal extends RuntimeVal {
  type: "native-fn";
  call: FunctionCall;
  properties: Map<string, RuntimeVal>;
}

export interface FunctionVal extends RuntimeVal {
    type: "function";
    name: string;
    parameters: string[];
    declarationEnv: Environment;
    body: Statement[];
    async: boolean;
}

export function MK_PROMISE(p: Promise<RuntimeVal>): PromiseVal {
    return { type: "promise", promise: p };
}

export function MK_NUMBER(n: number = 0): NumberVal {
  return { type: "number", value: n };
}

export function MK_NULL(): NullVal {
  return { type: "null", value: null };
}

export function MK_BOOL(b: boolean = true): BooleanVal {
  return { type: "boolean", value: b };
}

export function MK_STRING(s: string): StringVal {
    return { type: "string", value: s };
}

export function MK_NATIVE_FN(call: FunctionCall): NativeFnVal {
  return { type: "native-fn", call, properties: new Map() };
}

export function MK_OBJECT(properties: Map<string, RuntimeVal>): ObjectVal {
    return { type: "object", properties };
}

export function MK_ARRAY(elements: RuntimeVal[]): ArrayVal {
    return { type: "array", elements };
}

/**
 * Converts a NovaScript RuntimeVal back into a native JavaScript value.
 * This is used when passing NovaScript data to native functions or modules.
 */
export function runtimeToJsVal(val: RuntimeVal, fnWrapper?: (f: FunctionVal) => Function): any {
    if (val.underlyingValue !== undefined) return val.underlyingValue;

    switch (val.type) {
        case "number":
        case "string":
        case "boolean":
            return (val as any).value;
        case "null":
            return null;
        case "array":
            return (val as ArrayVal).elements.map(e => runtimeToJsVal(e, fnWrapper));
        case "object":
            const out: any = {};
            (val as ObjectVal).properties.forEach((v, k) => {
                out[k] = runtimeToJsVal(v, fnWrapper);
            });
            return out;
        case "promise":
            return (val as PromiseVal).promise;
        case "function":
            return fnWrapper ? fnWrapper(val as FunctionVal) : val;
        default:
            return val;
    }
}

/**
 * Converts a native JavaScript object/value into a NovaScript RuntimeVal.
 * This is crucial for bridging NPM modules into NovaScript.
 */
export function jsToRuntimeVal(jsObj: any, seen: Set<any> = new Set()): RuntimeVal {
    if (jsObj === null || jsObj === undefined) return MK_NULL();
    
    // Handle circular references
    if (typeof jsObj === "object" || typeof jsObj === "function") {
        if (seen.has(jsObj)) return MK_NULL();
        seen.add(jsObj);
    }

    if (typeof jsObj === "number") return MK_NUMBER(jsObj);
    if (typeof jsObj === "string") return MK_STRING(jsObj);
    if (typeof jsObj === "boolean") return MK_BOOL(jsObj);
    
    if (Array.isArray(jsObj)) {
        const arr = MK_ARRAY(jsObj.map(item => jsToRuntimeVal(item, seen)));
        arr.underlyingValue = jsObj;
        return arr;
    }
    
    const props = new Map<string, RuntimeVal>();
    
    const collectProperties = (obj: any) => {
        if (!obj || obj === Object.prototype || obj === Function.prototype) return;
        
        // Scan current level
        for (const key of Object.getOwnPropertyNames(obj)) {
            if (key === "prototype" || key === "length" || key === "name" || 
                key === "caller" || key === "arguments" || key === "constructor") continue;
            
            if (props.has(key)) continue;

            try {
                props.set(key, jsToRuntimeVal(obj[key], seen));
            } catch (e) {}
        }
        // Limit recursion depth for performance or rely on lazy resolution for prototype chain
    };

    if (typeof jsObj === "function") {
        const fn = MK_NATIVE_FN((args) => {
            const jsArgs = args.map(arg => runtimeToJsVal(arg));
            const result = jsObj(...jsArgs);
            return jsToRuntimeVal(result);
        });
        
        fn.underlyingValue = jsObj; // Store original for lazy resolution
        collectProperties(jsObj);
        fn.properties = props;
        return fn;
    }

    if (typeof jsObj === "object") {
        const obj = MK_OBJECT(props);
        obj.underlyingValue = jsObj; // Store original for lazy resolution
        collectProperties(jsObj);
        return obj;
    }
    
    return MK_NULL();
}

/**
 * Returns a plain string representation of a RuntimeVal without colors.
 */
export function plainStringify(val: RuntimeVal): string {
    switch (val.type) {
        case "string": return (val as any).value;
        case "number": return (val as any).value.toString();
        case "boolean": return (val as any).value.toString();
        case "null": return "null";
        case "promise": return "[Promise]";
        case "array": 
            return "[" + (val as any).elements.map((e: any) => plainStringify(e)).join(", ") + "]";
        case "object": 
            const props = Array.from((val as any).properties.entries())
                .map((entry: any) => {
                    const [k, v] = entry;
                    return `${k}: ${plainStringify(v as any)}`;
                })
                .join(", ");
            return "{ " + props + " }";
        case "function": return "[Function]";
        case "native-fn": return "[Native Function]";
        default: return JSON.stringify(val);
    }
}
